atlasdb-commons/src/test/java/com/palantir/common/collect/IterableViewTest.java

```diff
@@ -0,0 +1,114 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.common.collect;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.List;
+import java.util.Spliterator;
+
+import org.junit.Test;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.Iterables;
+
+public class IterableViewTest {
+
+    @Test
+    public void stream() {
+        assertThat(IterableView.of(ImmutableList.of("a", "b")).stream()).containsExactly("a", "b");
+        assertThat(IterableView.of(
+                Iterables.limit(Iterables.cycle(ImmutableSet.of("a", "b", "c")), 4)).stream())
+                .containsExactly("a", "b", "c", "a");
+        assertThat(IterableView.of(ImmutableList.of("a", "b")).transform(String::toUpperCase)
+                .stream())
+                .containsExactly("A", "B");
+    }
+
+    @Test
+    public void listSpliterator() {
+        Spliterator<String> spliterator = IterableView.of(ImmutableList.of("a", "b")).spliterator();
+        assertThat(spliterator.estimateSize()).isEqualTo(2);
+        int characteristics = spliterator.characteristics();
+        assertThat(characteristics).isNotEqualTo(0);
+        assertThat(characteristics & Spliterator.CONCURRENT).isEqualTo(0);
+        assertThat(characteristics & Spliterator.DISTINCT).isEqualTo(0);
+        assertThat(characteristics & Spliterator.IMMUTABLE).isEqualTo(Spliterator.IMMUTABLE);
+        assertThat(characteristics & Spliterator.NONNULL).isEqualTo(Spliterator.NONNULL);
+        assertThat(characteristics & Spliterator.SIZED).isEqualTo(Spliterator.SIZED);
+        assertThat(characteristics & Spliterator.SORTED).isEqualTo(0);
+        assertThat(characteristics & Spliterator.SUBSIZED).isEqualTo(Spliterator.SUBSIZED);
+    }
+
+    @Test
+    public void listPartitionSpliterator() {
+        IterableView<List<String>> view = IterableView.of(ImmutableList.of("a", "b", "c")).partition(1);
+        Spliterator<List<String>> spliterator = view.spliterator();
+        assertThat(spliterator.estimateSize()).isEqualTo(3);
+        int characteristics = spliterator.characteristics();
+        assertThat(characteristics).isNotEqualTo(0);
+        assertThat(characteristics & Spliterator.CONCURRENT).isEqualTo(0);
+        assertThat(characteristics & Spliterator.DISTINCT).isEqualTo(0);
+        assertThat(characteristics & Spliterator.IMMUTABLE).isEqualTo(0);
```

Comment:
These tests seem fragile to changes elsewhere. For example, it would be super reasonable for Guava's Iterables.partition to claim that it's non-null (always) and immutable (given immutable input), and for that to change in some later version of Guava.

I would propose that you test this with a single test, which simply validates that the spliterator characteristics are identical if you call delegate().spliterator() vs spliterator()?

Reply:
also, this is almost surely an oversight, but you aren't testing that the spliterators actually contain the right elements

Reply:
I'll trim them down to check that the returned spliterator has `SIZED` and `SUBSIZED` characteristics which are what we want to ensure that we can inherit sized optimizations that you don't get with a standard iterator based spliterator `java.util.Spliterators#spliteratorUnknownSize(java.util.Iterator<? extends T>, int)` from `java.lang.Iterable#spliterator`.

Reply:
I can't just assert that the spliterator's characteristics are identical to the delegate's as that doesn't always hold, e.g. for the transformed case of ImmutableList delegate, Guava does not currently propagate `IMMUTABLE` and `NONNULL` for a `com.google.common.collect.Lists$TransformingRandomAccessList`'s `spliterator`.

```
spliterator = {java.util.Spliterators$IteratorSpliterator@1192} 
 collection = {com.google.common.collect.Lists$TransformingRandomAccessList@1196}  size = 3
 it = {com.google.common.collect.Lists$TransformingRandomAccessList$1@1197} 
 characteristics = 16464 (0x4050)
 est = 3 (0x3)
 batch = 0 (0x0)
delegateSpliterator = {java.util.Spliterators$ArraySpliterator@1193} 
 array = {java.lang.Object[3]@1195} 
 index = 0 (0x0)
 fence = 3 (0x3)
 characteristics = 17744 (0x4550)
```